### ISSUE TITLE
Implement SharedState merge operation

### ISSUE THREAD

**Issue by madadam** [maidsafe/sn_routing#2118]
Implement the operation to merge one `SharedState` into another:

```rust
impl SharedState {
    // ...
    fn merge(&mut self, other: Self) -> Result<()>;
    // ...
}
```
Requirements:
- only modify `self` if `other` is valid (all entries of all containers must have valid BLS signatures against known keys)
- the operation must be commutative, associative and idempotent (semilattice/CRDT)

This operation will be used to promote nodes to elders (they will receive one or more `SharedState` instances from the other elders and will merge it into their own `SharedState` instance) and also to allow nodes that were temporarily disconnected or were experiencing network issues to catch up

Depends on #2123, #2125

**Comment by madadam** [comment]
`SharedState` was superseded by `Section` and `Network` structs, but both have a `merge` now. Closing.
